### PR TITLE
DrivAerML: 5L/384d — deeper model at wider dimension

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

5L/256d diverged in earlier runs, but 5L/384d may be stable. Wider layers provide more capacity per residual connection, potentially smoothing the optimization landscape. The jump from 4L→5L at 384d width (vs 256d) may give the model capacity to learn finer aerodynamic features without the instability seen at narrower widths. Adding one more layer to the current best (4L/384d, 5.73%) tests whether depth gains at this width are beneficial.

## Instructions

Run the following command exactly (one change from base: `--model-layers 5`):

```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 5 \
  --model-hidden-dim 384 \
  --model-heads 6 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --agent shinobu \
  --wandb-name "shinobu/drivaerml-5L384d" \
  --wandb-group "drivaerml-depth-at-384d"
```

**Single change from base:** `--model-layers 5` (all other hyperparameters identical to PR #2602)

**Note:** Monitor carefully for divergence (val metric spiking >20%). If it diverges in the first 20 epochs, stop early and report. If stable, let it run to the full 180-min timeout.

## Baseline

Current DrivAerML best (PR #2602):
- **val_primary/surface_rel_l2_pct:** 5.73% (epoch 144)
- **Baseline W&B run:** 7ogfs7ph (151 epochs, 180-min budget)
- **Architecture:** 4L/384d/6H + Fourier + no-EMA + AdamW lr=5e-4 + T_max=30
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 384 --model-heads 6 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`
- **External target:** <3.71% (AB-UPT)

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2684 because ADVISOR: Closing — 5L/384d superseded by throughput insight. 4L/320d (5.027%) gets 70% more training epochs..

**Comment by tcapelle** [comment]
STUDENT shinobu:

## Results

### Metrics

| Metric | Baseline (4L/384d, 151 ep) | This run (5L/384d, 266 ep) | Delta |
|--------|-----|-----|-------|
| **val_primary/surface_rel_l2_pct** | 5.73% (ep 144) | **4.73%** (ep 234) | **-1.00 pp** |
| **test_primary/surface_rel_l2_pct** | (see PR #2602) | **5.97%** | — |

### Best val_primary milestones

| Epoch | val_primary/surface_rel_l2_pct | Loss | LR |
|-------|-------------------------------|------|-----|
| 1 | 62.31% | 0.892 | 2.16e-5 |
| 5 | 41.55% | 0.323 | 3.75e-4 |
| 10 | 27.84% | 0.176 | 1.25e-4 |
| 20 | 22.00% | 0.093 | 1.25e-4 |
| 30 | 15.76% | 0.055 | 5.00e-4 |
| 75 | 8.48% | 0.020 | 0.0 |
| 100 | 7.50% | 0.014 | 5.00e-4 |
| 150 | 5.68% | 0.008 | 5.00e-4 |
| 179 | 5.30% | 0.007 | 2.16e-5 |
| 190 | 5.11% | 0.007 | 1.25e-4 |
| 211 | 5.02% | 0.006 | 2.16e-5 |
| 218 | 4.91% | 0.006 | 5.46e-6 |
| **234** | **4.73%** | **0.006** | 4.77e-5 |

### Command

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 5 \
  --model-hidden-dim 384 \
  --model-heads 6 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --agent shinobu \
  --wandb-name "shinobu/drivaerml-5L384d" \
  --wandb-group "drivaerml-depth-at-384d"
```

### Details

- **W&B run:** `4to5rag7` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/4to5rag7))
- **Peak memory:** ~6.9 GB GPU VRAM (GPU 0, NVIDIA RTX PRO 6000 Blackwell)
- **Epochs completed:** 266 (hit 180-min timeout)
- **Time per epoch:** ~41 seconds
- **Architecture:** 5L/384d/6H + Fourier + no-EMA + AdamW lr=5e-4 + T_max=30

### What happened

**The hypothesis is confirmed — 5L/384d beats 4L/384d.** Best val dropped from 5.73% to 4.73%, a 1.0 pp improvement. Crucially, 5L/384d showed **no divergence** — unlike earlier 5L/256d attempts that diverged. The wider 384d dimension stabilizes the extra depth.

Key observations:

1. **No divergence at 5L.** Val metrics decreased steadily throughout training with normal cosine oscillations. The concern about instability at 5 layers was unfounded at 384d width.

2. **Convergence was steady but slow.** The model crossed the 5.73% baseline around epoch 150 and continued improving to 4.73% at epoch 234. The final improvement from 5.73% → 4.73% took ~100 epochs.

3. **Test/val gap.** Best val was 4.73% but test was 5.97% — a 1.24 pp gap. Note the test eval ran at epoch 266 (not the best-val epoch 234), and val was 5.17% at that point. The gap suggests some val-specific fitting, worth investigating.

4. **Still improving at termination.** Best val was set at epoch 234 of 266, and the model was still making occasional new bests. More training time could push below 4.5%.

5. **Approaching the external target.** The external AB-UPT target is <3.71%. At 4.73% val, we're within 1 pp — continued depth/width scaling may close this gap.

### Suggested follow-ups

1. **6L/384d** — if 5L beats 4L without divergence, does 6L push further? The 384d width seems sufficient to stabilize deep models.
2. **5L/512d** — wider at 5L could combine depth and width gains.
3. **5L/384d + T_max=5** — my earlier T_max=5 experiment showed dramatic gains from rapid cycling. Combining with the deeper 5L model could compound both benefits.
4. **Investigate test/val gap** — the 1.24 pp gap at final eval is concerning. May need eval-time point sampling changes or checkpoint selection at best-val epoch.
